### PR TITLE
Add station in treeview

### PR DESCRIPTION
**Pull request by JasonLG1979**
Adds the ability to add stations in the treeview based on the currently selected song/artist.

### CODE REVIEW

File: pithos/StationsDialog.py [hunk 1]
```diff
@@ -196,72 +195,46 @@ def refresh_stations(self, widget):
...
+            description = html.escape(result.stationName)
+        user_data = description, result.resultType
+        logging.info("in add_station_cb {} {}".format(result, response))
         if response == Gtk.ResponseType.OK:
-            self.worker_run("add_station_by_music_id", (self.result.musicId,),
-                            self.station_added, "Creating station...")
+            self.worker_run(
+                "add_station_by_music_id",
+                (result.musicId,),
+                self.station_added,
```
Comment (TingPing) [inline pithos/StationsDialog.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/StationsDialog.py [hunk 2]
```diff
@@ -196,72 +195,46 @@ def refresh_stations(self, widget):
...
-        dialog.connect('response', on_response)
-        dialog.show()
-
-    def station_added(self, station):
+    def station_added(self, station, user_data):
+        description, music_type = user_data
         for existing_station in self.model:
-            if existing_station[0].id == station.id:
-                self.station_already_exists(existing_station[0])
+            if existing_station[0].id == station.id:
```
Comment (TingPing) [inline pithos/StationsDialog.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pithos.py [hunk 3]
```diff
@@ -910,6 +918,66 @@ def station_changed(self, station, reconnecting=False):
...
+            listen_now = _('Would you like to listen to it now?')
+            message = '{}\n{} "{}" {} {} {}.\n{}'.format(
+                sub_title,
+                your_station,
+                station.name,
+                already_contains,
+                seed,
+                description,
+                listen_now,
+           )
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 4]
```diff
@@ -1213,6 +1281,16 @@ def on_menuitem_bookmark_song(self, widget):
     def on_menuitem_bookmark_artist(self, widget):
         self.bookmark_song_artist(self.selected_song())
 
+    @GtkTemplate.Callback
+    def on_menuitem_create_artist_station(self, widget):
+        user_date = 'artist', html.escape(self.selected_song().artist)
+        self.worker_run('add_station_by_track_token', (self.selected_song().trackToken, 'artist'), self.station_added, user_data=user_date)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E501] line too long (139 > 120 characters)

File: pithos/pithos.py [hunk 5]
```diff
@@ -1213,6 +1281,16 @@ def on_menuitem_bookmark_song(self, widget):
...
+
+    @GtkTemplate.Callback
+    def on_menuitem_create_song_station(self, widget):
+        user_date = 'song', '{} by {}'.format(html.escape(self.selected_song().title), html.escape(self.selected_song().artist))
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E501] line too long (128 > 120 characters)

File: pithos/pithos.py [hunk 6]
```diff
@@ -1213,6 +1281,16 @@ def on_menuitem_bookmark_song(self, widget):
...
+        self.worker_run('add_station_by_track_token', (self.selected_song().trackToken, 'song'), self.station_added, user_data=user_date)
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E501] line too long (137 > 120 characters)

File: pithos/StationsDialog.py [hunk 7]
```diff
@@ -196,72 +195,47 @@ def refresh_stations(self, widget):
...
```
Comment (TingPing) [inline pithos/StationsDialog.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/pithos.py [hunk 8]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+            message = '{}\n"{}"{} {} {}'.format(sub_title, station.name, message, seed, description)
+
+        else:
+            button_type = Gtk.ButtonsType.YES_NO
+            your_station = _('Your Station')
+            already_contains = _('already contains the')
+            listen_now = _('Would you like to listen to it now?')
+            message = (
+                '{}\n{} "{}" {} {} {}.\n{}'.format(
+                sub_title,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 9]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+                your_station,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 10]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+                station.name,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 11]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+                already_contains,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 12]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+                seed,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 13]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+                description,
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

File: pithos/pithos.py [hunk 14]
```diff
@@ -910,6 +918,67 @@ def station_changed(self, station, reconnecting=False):
...
+                listen_now,
+           ))
```
Comment (TingPing) [inline pithos/pithos.py]: __[flake8]__
* [E122] continuation line missing indentation or outdented

**Comment by JasonLG1979** [comment]
@TingPing This and the mediakeys fix and then 1.3.1?

File: pithos/pithos.py [hunk 15]
```diff
@@ -910,6 +918,58 @@ def station_changed(self, station, reconnecting=False):
...
+
+        if station is self.current_station:
+            button_type = Gtk.ButtonsType.OK
+            message = _(', the Station you are currently listening to already contains the')
+            message = '{}\n"{}"{} {} {}'.format(sub_title, station.name, message, seed, description)
+
+        else:
+            button_type = Gtk.ButtonsType.YES_NO
+            a = _('Your Station')
+            b = _('already contains the')
```
Comment (TingPing) [inline pithos/pithos.py]: These partial snippets of sentences aren't really translatable, could you perhaps break them up into a few complete sentences?

**Comment by TingPing** [comment]
Great job on the UX, you really handled each edge case really well.

**Comment by JasonLG1979** [comment]
Thank!!!, I'm really trying to avoid just throwing "Pandora Error" at a user when something doesn't go as expected.

File: pithos/pithos.py (hunk 15 above)
Comment (JasonLG1979) [inline pithos/pithos.py]: I'm not sure how translations work. There are several variables that shouldn't be translated in the middle of bits that should be translated.

Comment (TingPing) [inline pithos/pithos.py]: Whatever is with in a `_()` is picked up for translations and the function just returns a translated string. So you can break up strings how ever you want and only translate parts of it but it should at least be understandable to a translator.

I do realize we don't actually have translations atm but I'd like for that to change.

**Comment by JasonLG1979** [comment]
Does that work? now they are complete sentences and .format refs the untranslatable bits by index.

**Comment by JasonLG1979** [comment]
The translator can then decide where they want to put the bit in format by moving around the "{index}"

**Comment by JasonLG1979** [comment]
> I do realize we don't actually have translations atm but I'd like for that to change.

At some point someone(prob me) will have to go through everything and find all the translatable strings.

**Pull request merged**